File: src/swingjs/JSMouse.java
/* $RCSfile$
 * $Author: hansonr $
 * $Date: 2012-09-11 19:29:26 -0500 (Tue, 11 Sep 2012) $
 * $Revision: 17556 $
 *
 * Copyright (C) 2002-2005  The Jmol Development Team
 *
 * Contact: jmol-developers@lists.sf.net
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 */
package swingjs;

import java.awt.event.InputEvent;

import jsjava.awt.Component;
import jsjava.awt.Event;
import jsjava.awt.Toolkit;
import jsjava.awt.event.MouseEvent;
import jsjava.awt.event.MouseWheelEvent;

/**
 * JavaScript interface from JmolJSmol.js via handleOldJvm10Event (for now)
 * 
 * J2SRequireImport is needed because we want to allow JavaScript access to
 * java.awt.Event constant names
 * 
 */

public class JSMouse {

	private JSFrameViewer viewer;
	private Object jqevent;

	public JSMouse(JSFrameViewer v) {
		viewer = v;
	}

	public boolean processEvent(int id, int x, int y, int modifiers, long time, Object jqevent) {
		this.jqevent = jqevent;
		if (id != -1 && id != Event.MOUSE_MOVE)
			modifiers = applyLeftMouse(modifiers);
		switch (id) {
		case -1: // JavaScript
			wheeled(time, x, modifiers);
			break;
		case Event.MOUSE_DOWN:
			xWhenPressed = x;
			yWhenPressed = y;
			modifiersWhenPressed10 = modifiers;
			pressed(time, x, y, modifiers, false);
			break;
		case Event.MOUSE_DRAG:
			dragged(time, x, y, modifiers);
			break;
		case Event.MOUSE_ENTER:
			entry(time, x, y, false);
			break;
		case Event.MOUSE_EXIT:
			entry(time, x, y, true);
			break;
		case Event.MOUSE_MOVE:
			moved(time, x, y, modifiers);
			break;
		case Event.MOUSE_UP:
			released(time, x, y, modifiers);
			// simulate a mouseClicked event for us
			if (x == xWhenPressed && y == yWhenPressed
					&& modifiers == modifiersWhenPressed10) {
				// the underlying code will turn this into dbl clicks for us
				clicked(time, x, y, modifiers, 1);
			}
			break;
		default:
			return false;
		}
		return true;
	}

	/**
	 * 
	 * called by JSmol as processTwoPointGesture(canvas.touches);
	 * 
	 * @param touches
	 *          [[finger1 touches],[finger2 touches]] where finger touches are
	 *          [[x0,y0],[x1,y1],[x2,y2],...]
	 * 
	 */
	public void processTwoPointGesture(float[][][] touches) {
		getMouse2().processTwoPointGesture(touches);
	}

	private JSMouse2 mouse2;
	private JSMouse2 getMouse2() {
		return (mouse2 == null ? (mouse2 = (JSMouse2) JSToolkit.getInstance("swingjs.JSMouse2")).set(this) : mouse2);
	}

	public static final int MOUSE_LEFT = 16; // MouseEvent.BUTTON1_MASK
	public static final int MOUSE_MIDDLE = 8; // Event.ALT_MASK;  MouseEvent.BUTTON2_MASK
	public static final int MOUSE_RIGHT = 4; // Event.META_MASK;  MouseEvent.BUTTON3_MASK
	public static final int MOUSE_WHEEL = 32;

	public static final int EXTENDED_MASK = 0x3FC0; // ^(InputEvent.HIGH_MODIFIERS | JDK_1_3_MODIFIERS)  
	public final static int MAC_COMMAND = MOUSE_LEFT | MOUSE_RIGHT;
	public final static int BUTTON_MASK = MOUSE_LEFT | MOUSE_MIDDLE | MOUSE_RIGHT;
	
	void translateXYBy(int deltaX, int deltaY) {
		// TODO Auto-generated method stub

	}

	public void mouseClicked(MouseEvent e) {
		clicked(e.getWhen(), e.getX(), e.getY(), e.getModifiers(),
				e.getClickCount());
	}

	public void mouseEntered(MouseEvent e) {
		entry(e.getWhen(), e.getX(), e.getY(), false);
	}

	public void mouseExited(MouseEvent e) {
		entry(e.getWhen(), e.getX(), e.getY(), true);
	}

	public void mousePressed(MouseEvent e) {
		pressed(e.getWhen(), e.getX(), e.getY(), e.getModifiers(),
				e.isPopupTrigger());
	}

	public void mouseReleased(MouseEvent e) {
		released(e.getWhen(), e.getX(), e.getY(), e.getModifiers());
	}

	public void mouseDragged(MouseEvent e) {
		int modifiers = e.getModifiers();
		/****************************************************************
		 * Netscape 4.* Win32 has a problem with mouseDragged if you left-drag then
		 * none of the modifiers are selected we will try to fix that here
		 ****************************************************************/
		if ((modifiers & BUTTON_MASK) == 0)
			modifiers |= MOUSE_LEFT;
		/****************************************************************/
		dragged(e.getWhen(), e.getX(), e.getY(), modifiers);
	}

	public void mouseMoved(MouseEvent e) {
		moved(e.getWhen(), e.getX(), e.getY(), e.getModifiers());
	}

	public void mouseWheelMoved(MouseWheelEvent e) {
		e.consume();
		wheeled(e.getWhen(), e.getWheelRotation(), e.getModifiers());
	}

	private void entry(long time, int x, int y, boolean isExit) {
		wheeling = false;
		mouseEnterExit(time, x, y, isExit);
	}

	/**
	 * 
	 * @param time
	 * @param x
	 * @param y
	 * @param modifiers
	 * @param clickCount
	 */
	private void clicked(long time, int x, int y, int modifiers, int clickCount) {
		// clearKeyBuffer();
		// clickedCount is not reliable on some platforms
		// so we will just deal with it ourselves
		mouseAction(MouseEvent.MOUSE_CLICKED, time, x, y, 1, modifiers);
	}

	private boolean isMouseDown; // Macintosh may not recognize CTRL-SHIFT-LEFT as
																// drag, only move
	private boolean wheeling;

	private void moved(long time, int x, int y, int modifiers) {
		// clearKeyBuffer();
		if (isMouseDown)
			mouseAction(MouseEvent.MOUSE_DRAGGED, time, x, y, 0,
					applyLeftMouse(modifiers));
		else
			mouseAction(MouseEvent.MOUSE_MOVED, time, x, y, 0, modifiers);
	}

	void wheeled(long time, int rotation, int modifiers) {
		// clearKeyBuffer();
		wheeling = true;
		mouseAction(MouseEvent.MOUSE_WHEEL, time, 0, rotation, 0, modifiers
				& ~BUTTON_MASK | MOUSE_WHEEL);
	}

	/**
	 * 
	 * @param time
	 * @param x
	 * @param y
	 * @param modifiers
	 * @param isPopupTrigger
	 */
	private void pressed(long time, int x, int y, int modifiers,
			boolean isPopupTrigger) {
		// clearKeyBuffer();
		isMouseDown = true;
		wheeling = false;
		mouseAction(MouseEvent.MOUSE_PRESSED, time, x, y, 0, modifiers);
	}

	private void released(long time, int x, int y, int modifiers) {
		isMouseDown = false;
		wheeling = false;
		mouseAction(MouseEvent.MOUSE_RELEASED, time, x, y, 0, modifiers);
	}

	private void dragged(long time, int x, int y, int modifiers) {
		if (wheeling)
			return;
		if ((modifiers & MAC_COMMAND) == MAC_COMMAND)
			modifiers = modifiers & ~MOUSE_RIGHT | Event.CTRL_MASK;
		mouseAction(MouseEvent.MOUSE_DRAGGED, time, x, y, 0, modifiers);
	}

	private static int applyLeftMouse(int modifiers) {
		// if neither BUTTON2 or BUTTON3 then it must be BUTTON1
		return ((modifiers & BUTTON_MASK) == 0) ? (modifiers | MOUSE_LEFT)
				: modifiers;
	}

	private int xWhenPressed, yWhenPressed, modifiersWhenPressed10;

	private int getButton(int modifiers) {
		switch (modifiers & BUTTON_MASK) {
		case MOUSE_LEFT:
			return MouseEvent.BUTTON1;
		case MOUSE_MIDDLE:
			return MouseEvent.BUTTON2;
		case MOUSE_RIGHT:
			return MouseEvent.BUTTON3;
		default:
			return MouseEvent.NOBUTTON;
		}
	}

	private void mouseEnterExit(long time, int x, int y, boolean isExit) {
		// nothing here?
	}

	private void mouseAction(int id, long time, int x, int y,
			int count, int modifiers) {
	  boolean popupTrigger = 
	      ( (modifiers & EXTENDED_MASK) == 
	          (JSToolkit.isMac ? InputEvent.CTRL_DOWN_MASK | InputEvent.BUTTON1_DOWN_MASK
	          : InputEvent.BUTTON3_DOWN_MASK)
	       );
		int button = getButton(modifiers);
		Component source = viewer.top; // may be a JFrame
		MouseEvent e = new MouseEvent(source, id, time, modifiers, x, y, x, y, count, popupTrigger, button);
		byte[] bdata = new byte[0];
		Object jqevent = this.jqevent;
		/**
		 * @j2sNative
		 *  bdata.jqevent = this.jqevent; 
		 */
		{
			System.out.println(jqevent);
		}
		e.setBData(e, bdata);
		Toolkit.getEventQueue().postEvent(e);
	}


}

File: site/swingjs/j2s/swingjs/JSMouse.js
Clazz.declarePackage ("swingjs");
Clazz.load (null, "swingjs.JSMouse", ["JU.V3", "java.awt.Toolkit", "java.awt.event.MouseEvent"], function () {
c$ = Clazz.decorateAsClass (function () {
this.ap = null;
this.isMouseDown = false;
this.wheeling = false;
this.xWhenPressed = 0;
this.yWhenPressed = 0;
this.modifiersWhenPressed10 = 0;
Clazz.instantialize (this, arguments);
}, swingjs, "JSMouse");
Clazz.makeConstructor (c$, 
function (ap) {
this.ap = ap;
}, "swingjs.JSAppletPanel");
Clazz.defineMethod (c$, "processEvent", 
function (id, x, y, modifiers, time) {
if (id != -1) modifiers = swingjs.JSMouse.applyLeftMouse (modifiers);
switch (id) {
case -1:
this.wheeled (time, x, modifiers);
break;
case 501:
this.xWhenPressed = x;
this.yWhenPressed = y;
this.modifiersWhenPressed10 = modifiers;
this.pressed (time, x, y, modifiers, false);
break;
case 506:
this.dragged (time, x, y, modifiers);
break;
case 504:
this.entry (time, x, y, false);
break;
case 505:
this.entry (time, x, y, true);
break;
case 503:
this.moved (time, x, y, modifiers);
break;
case 502:
this.released (time, x, y, modifiers);
if (x == this.xWhenPressed && y == this.yWhenPressed && modifiers == this.modifiersWhenPressed10) {
this.clicked (time, x, y, modifiers, 1);
}break;
default:
return false;
}
return true;
}, "~N,~N,~N,~N,~N");
Clazz.defineMethod (c$, "processTwoPointGesture", 
function (touches) {
if (touches[0].length < 2) return;
var t1 = touches[0];
var t2 = touches[1];
var t1first = t1[0];
var t1last = t1[t2.length - 1];
var x1first = t1first[0];
var x1last = t1last[0];
var dx1 = x1last - x1first;
var y1first = t1first[1];
var y1last = t1last[1];
var dy1 = y1last - y1first;
var v1 = JU.V3.new3 (dx1, dy1, 0);
var d1 = v1.length ();
var t2first = t2[0];
var t2last = t2[t2.length - 1];
var x2first = t2first[0];
var x2last = t2last[0];
var dx2 = x2last - x2first;
var y2first = t2first[1];
var y2last = t2last[1];
var dy2 = y2last - y2first;
var v2 = JU.V3.new3 (dx2, dy2, 0);
var d2 = v2.length ();
if (d1 < 1 || d2 < 1) return;
v1.normalize ();
v2.normalize ();
var cos12 = (v1.dot (v2));
if (cos12 > 0.8) {
var deltaX = Clazz.floatToInt (x1last - t1[t1.length - 2][0]);
var deltaY = Clazz.floatToInt (y1last - t1[t1.length - 2][1]);
this.translateXYBy (deltaX, deltaY);
} else if (cos12 < -0.8) {
v1 = JU.V3.new3 (x2first - x1first, y2first - y1first, 0);
v2 = JU.V3.new3 (x2last - x1last, y2last - y1last, 0);
var dx = v2.length () - v1.length ();
this.wheeled (System.currentTimeMillis (), dx < 0 ? -1 : 1, 32);
}}, "~A");
Clazz.defineMethod (c$, "translateXYBy", 
 function (deltaX, deltaY) {
}, "~N,~N");
Clazz.defineMethod (c$, "mouseClicked", 
function (e) {
this.clicked (e.getWhen (), e.getX (), e.getY (), e.getModifiers (), e.getClickCount ());
}, "java.awt.event.MouseEvent");
Clazz.defineMethod (c$, "mouseEntered", 
function (e) {
this.entry (e.getWhen (), e.getX (), e.getY (), false);
}, "java.awt.event.MouseEvent");
Clazz.defineMethod (c$, "mouseExited", 
function (e) {
this.entry (e.getWhen (), e.getX (), e.getY (), true);
}, "java.awt.event.MouseEvent");
Clazz.defineMethod (c$, "mousePressed", 
function (e) {
this.pressed (e.getWhen (), e.getX (), e.getY (), e.getModifiers (), e.isPopupTrigger ());
}, "java.awt.event.MouseEvent");
Clazz.defineMethod (c$, "mouseReleased", 
function (e) {
this.released (e.getWhen (), e.getX (), e.getY (), e.getModifiers ());
}, "java.awt.event.MouseEvent");
Clazz.defineMethod (c$, "mouseDragged", 
function (e) {
var modifiers = e.getModifiers ();
if ((modifiers & 28) == 0) modifiers |= 16;
this.dragged (e.getWhen (), e.getX (), e.getY (), modifiers);
}, "java.awt.event.MouseEvent");
Clazz.defineMethod (c$, "mouseMoved", 
function (e) {
this.moved (e.getWhen (), e.getX (), e.getY (), e.getModifiers ());
}, "java.awt.event.MouseEvent");
Clazz.defineMethod (c$, "mouseWheelMoved", 
function (e) {
e.consume ();
this.wheeled (e.getWhen (), e.getWheelRotation (), e.getModifiers ());
}, "java.awt.event.MouseWheelEvent");
Clazz.defineMethod (c$, "entry", 
 function (time, x, y, isExit) {
this.wheeling = false;
this.mouseEnterExit (time, x, y, isExit);
}, "~N,~N,~N,~B");
Clazz.defineMethod (c$, "clicked", 
 function (time, x, y, modifiers, clickCount) {
this.mouseAction (500, time, x, y, 1, modifiers);
}, "~N,~N,~N,~N,~N");
Clazz.defineMethod (c$, "moved", 
 function (time, x, y, modifiers) {
if (this.isMouseDown) this.mouseAction (506, time, x, y, 0, swingjs.JSMouse.applyLeftMouse (modifiers));
 else this.mouseAction (503, time, x, y, 0, modifiers);
}, "~N,~N,~N,~N");
Clazz.defineMethod (c$, "wheeled", 
 function (time, rotation, modifiers) {
this.wheeling = true;
this.mouseAction (507, time, 0, rotation, 0, modifiers & -29 | 32);
}, "~N,~N,~N");
Clazz.defineMethod (c$, "pressed", 
 function (time, x, y, modifiers, isPopupTrigger) {
this.isMouseDown = true;
this.wheeling = false;
this.mouseAction (501, time, x, y, 0, modifiers);
}, "~N,~N,~N,~N,~B");
Clazz.defineMethod (c$, "released", 
 function (time, x, y, modifiers) {
this.isMouseDown = false;
this.wheeling = false;
this.mouseAction (502, time, x, y, 0, modifiers);
}, "~N,~N,~N,~N");
Clazz.defineMethod (c$, "dragged", 
 function (time, x, y, modifiers) {
if (this.wheeling) return;
if ((modifiers & 20) == 20) modifiers = modifiers & -5 | 2;
this.mouseAction (506, time, x, y, 0, modifiers);
}, "~N,~N,~N,~N");
c$.applyLeftMouse = Clazz.defineMethod (c$, "applyLeftMouse", 
 function (modifiers) {
return ((modifiers & 28) == 0) ? (modifiers | 16) : modifiers;
}, "~N");
Clazz.defineMethod (c$, "getButton", 
 function (modifiers) {
switch (modifiers & 28) {
case 16:
return 1;
case 8:
return 2;
case 4:
return 3;
default:
return 0;
}
}, "~N");
Clazz.defineMethod (c$, "mouseEnterExit", 
 function (time, x, y, isExit) {
}, "~N,~N,~N,~B");
Clazz.defineMethod (c$, "mouseAction", 
 function (id, time, x, y, count, modifiers) {
var popupTrigger = false;
var button = this.getButton (modifiers);
var source = this.ap.applet;
var e =  new java.awt.event.MouseEvent (source, id, time, modifiers, x, y, x, y, count, popupTrigger, button);
java.awt.Toolkit.getEventQueue ().postEvent (e);
}, "~N,~N,~N,~N,~N,~N");
Clazz.defineStatics (c$,
"MOUSE_LEFT", 16,
"MOUSE_MIDDLE", 8,
"MOUSE_RIGHT", 4,
"MOUSE_WHEEL", 32,
"MAC_COMMAND", 20,
"BUTTON_MASK", 28);
});
